### PR TITLE
C++: SSA flow through fields and imprecise defs

### PR DESCRIPTION
**Pull request by dave-bartolomeo**
The diffs in the SSAConstruction.qll are pretty extensive. While the overall approach is similar to what was there before, you're probably better off just looking at the new implementation without trying to worry about line-by-line differences from the old implementation.

If you look at the diffs to the test expectations commit-by-commit, you'll see how the newly-generated SSA differs from the previous one. Primarily, you'll see uses getting hooked up to exact definitions if possible, instead of always being hooked up to the result of a `Chi`. You'll also see `Phi` instructions inserted for fields, instead of just for entire virtual variables.

Most of the changes are described in the markdown files that I've added alongside the code.

### CODE REVIEW

**Comment by jbj** [comment]
I looked at the changes to `Operand` and test cases, and they LGTM. Some of the changes don't seem to show up in test results, so maybe we need more tests. The new def-use implementation is going to be hard to review.

**Comment by dave-bartolomeo** [comment]
I've pushed new changes, mostly to SSAConstruction.qll. SSA construction is now grouped into separate modules for Phi Insertion and Use/Def connection, which will hopefully make it easier to review. Each module has an overview comment that outlines the general approach used. Also, I believe that all predicates in SSA construction now have QLDoc.

File: cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/internal/AliasedSSA.qll [hunk 1]
```diff
@@ -1,313 +1,278 @@
...
-  }
-}
-
-private predicate hasResultMemoryAccess(Instruction instr, IRVariable var, IntValue startBitOffset,
+private predicate hasResultMemoryAccess(Instruction instr, IRVariable var, Type type, IntValue startBitOffset,
     IntValue endBitOffset) {
   resultPointsTo(instr.getResultAddressOperand().getDefinitionInstruction(), var, startBitOffset) and
+  type = instr.getResultType() and
   if exists(instr.getResultSize()) then
     endBitOffset = Ints::add(startBitOffset, Ints::mul(instr.getResultSize(), 8))
```
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/internal/AliasedSSA.qll]: Why `Ints::mul` and not just `*`? Also below.

File: cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/internal/SSAConstruction.qll [hunk 2]
```diff
@@ -372,185 +359,418 @@ cached private module Cached {
...
-    variableLiveOnEntryToBlock(vvar, block.getAFeasibleSuccessor())
+  predicate definitionLiveOnExitFromBlock(Alias::MemoryLocation defLocation, OldBlock block) {
+    definitionLiveOnEntryToBlock(defLocation, block.getAFeasibleSuccessor())
   }
+}
 
+private import DefUse
+
+/**
+ * Module containing the predicates that connect uses to their reaching definition. The reaching definitios are computed
```
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/internal/SSAConstruction.qll]: Typo: definitios

File: cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/internal/AliasedSSA.qll [hunk 3]
```diff
@@ -1,313 +1,278 @@
...
-  result = TVariableMemoryAccess(var, startBitOffset, endBitOffset)
-}
+  TUnknownMemoryLocation(FunctionIR funcIR) or
+  TUnknownVirtualVariable(FunctionIR funcIR)
 
-class MemoryAccess extends TMemoryAccess {
-  string toString() {
-    none()
-  }
+abstract class MemoryLocation extends TMemoryLocation {
```
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/internal/AliasedSSA.qll]: Please add QLDoc for `MemoryLocation`, explaining how it relates to variables.

**Comment by dave-bartolomeo** [comment]
I've added a .md file that gives an overview of the whole SSA construction process, from alias analysis to the memory model to SSA itself. That doc is missing the details on the actual connection of defs to uses, but I wanted to get the other parts of it into the PR to give the necessary background to reviewers.

File: cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/internal/AliasedSSA.qll (hunk 1 above)
Comment (dave-bartolomeo) [reply cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/internal/AliasedSSA.qll]: I think these should probably remain. The `startBitOffset` bound by `resultPointsTo()` could be the result of arbitrary pointer arithmetic. Overflow is unlikely, but possible.

File: cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/internal/AliasedSSA.qll [hunk 4]
```diff
@@ -1,313 +1,278 @@
...
+    else
+      result = TVariableMemoryLocation(var, var.getType(), 0, var.getType().getSize() * 8)
+  }
+
+  /**
+   * Holds if this memory location covers the entire variable.
+   */
+  final predicate coversEntireVariable() {
+    startBitOffset = 0 and
+    Ints::isEQ(endBitOffset, Ints::mul(var.getType().getSize(), 8))
```
[the remark this answers is not in the document]
Comment (dave-bartolomeo) [reply cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/internal/AliasedSSA.qll]: Fixed.

File: cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/internal/SSAConstruction.qll [hunk 5]
```diff
@@ -372,185 +359,418 @@ cached private module Cached {
...
+private import PhiInsertion
+
+/**
+ * Module to handle insertion of `Phi` instructions at the correct blocks. We insert a `Phi` instruction at the
+ * beginning of a block for a given location when that block is on the dominance frontier of a definition of the
+ * location and there is a use of that location reachable from that block without an intervening definition of the
+ * location.
+ * Within the approach outlined above, we treat a location slightly differently depending on whether or not it is a
+ * virtual variable. For a virtual variable, 
+ * location into the beginning of a block
```
[the remark this answers is not in the document]
Comment (dave-bartolomeo) [reply cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/internal/SSAConstruction.qll]: Fixed

File: cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/internal/SSAConstruction.qll (hunk 2 above)
Comment (dave-bartolomeo) [reply cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/internal/SSAConstruction.qll]: Fixed

File: cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/internal/AliasedSSA.qll (hunk 3 above)
Comment (dave-bartolomeo) [reply cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/internal/AliasedSSA.qll]: Done

File: cpp/ql/src/semmle/code/cpp/ir/implementation/SSA.md [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+and always accessed in their entirety via their declared type. There is one `MemoryLocation` for
+each unescaped `IRVariable`, and each `MemoryLocation` is its own `VirtualVariable`. The overlap
+relationship is simple: Each `MemoryLocation` exactly overlaps itself, and does not overlap any
+other `MemoryLocation`.
+
+#### Aliased SSA
+The current memory model used to construct Aliased SSA models every memory access. There are two
+kinds of `MemoryLocation`:
+- `VariableMemoryAccess` represents an access to a known `IRVariable` with a specific type, at a bit
+offset that may or may not be a known constant. `VariableMemoryAccess` represents any access to a
```
[the remark this answers is not in the document]
Comment (dave-bartolomeo) [reply cpp/ql/src/semmle/code/cpp/ir/implementation/SSA.md]: What's odd is that I wrote this .md file a full week after I switched to the new name. Fixed.

File: cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/internal/AliasedSSA.qll (hunk 1 above)
Comment (jbj) [reply cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/internal/AliasedSSA.qll]: I agree that `Ints::add` should remain, but my question was about `Ints::mul`.

File: cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/internal/SSAConstruction.qll [hunk 7]
```diff
@@ -372,185 +359,425 @@ cached private module Cached {
...
 
-  private predicate definitionReachesUseWithinBlock(Alias::VirtualVariable vvar, OldBlock defBlock,
-      int defRank, OldBlock useBlock, int useRank) {
-    defBlock = useBlock and
-    hasDefinitionAtRank(vvar, defBlock, defRank, _) and
-    hasUseAtRank(vvar, useBlock, useRank, _) and
-    definitionReachesRank(vvar, defBlock, defRank, useRank)
+  /**
+   * Holds if the specified `useLocation` is live on entry to `block`. This holds if there is a use of `useLocation`
+   * that is reachable from the start of `block` without passing through a definition that overlaps `useLocation`.
```
Comment (rdmarsh2) [inline cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/internal/SSAConstruction.qll]: Should this be "without passing through a definition that totally overlaps useLocation"? If not, can add a clarification about how this interacts with Chi nodes?

**Comment by dave-bartolomeo** [comment]
Initial perf info on ChakraCore: Running constant_func.ql, which consumes aliased SSA, the total slowdown was <3%, which is actually much faster than I expected. However, it doesn't seem to be improving the precision of the use/def connections as much as I expected it to. I'm investigating that now.

Wireshark numbers to come.

**Comment by jbj** [comment]
The mergeback #1082 just went in, so I'll try to retarget this PR to `master`.

**Comment by jbj** [comment]
Can you summarise what has changed since I last looked at this?

**Comment by dave-bartolomeo** [comment]
There have been few real changes. Mostly, it's a rebase, plus updating test expectations for IR construction improvements that have been merged since the previous review. The dataflow test expectation updates do show a couple concrete examples of where we get better results than the AST-based dataflow because we can model field accesses more precisely.

**Comment by dave-bartolomeo** [comment]
OK, now there's been one significant change: in an IR dump, an operand that is not an exact use of its definition instruction (e.g. load of a field defined by a store of the entire struct) now has a "~" prefix. I had implemented this earlier while working on this PR, but only enabled it just now to avoid polluting the IR diffs caused by real SSA changes.

**Comment by jbj** [comment]
There are still two unaddressed comments from earlier reviews by @rdmarsh2 and me. Also, you promised performance testing on Wireshark. I think this kind of change ought to be benchmarked on a handful of big snapshots since it may be sensitive to specific patterns that confuse the alias analysis in very large functions.

**Comment by dave-bartolomeo** [comment]
Latest run on Wireshark shows an 8% overall slowdown to evaluate all the way to aliased SSA. This is more than the 2.5% slowdown in ChakraCore, but still consistent with Wireshark's reputation as a stress test for this sort of change.

File: cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/internal/AliasedSSA.qll (hunk 1 above)
Comment (dave-bartolomeo) [reply cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/internal/AliasedSSA.qll]: Overflow in the `mul` calculation is even less likely, but still theoretically possible if the result of the instruction is a very large array type. Is there likely to be a significant performance cost from this use of `mul`? If so, it's probably OK to remove it, but if there's not likely to be much performance difference, I'd prefer to keep it for (pedantic) correctness.

File: cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/internal/SSAConstruction.qll (hunk 7 above)
Comment (dave-bartolomeo) [reply cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/internal/SSAConstruction.qll]: Clarification added

**Comment by dave-bartolomeo** [comment]
@rdmarsh2 I believe this is ready for final approval and merge, if the 8% slowdown on Wireshark seems reasonable.

**Comment by rdmarsh2** [comment]
Have you looked at the change in the most expensive predicates with and without this PR? I think we can merge with 8%, but we shouldn't introduce low-hanging fruit for performance improvements.

File: cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/internal/AliasedSSA.qll (hunk 1 above)
Comment (jbj) [reply cpp/ql/src/semmle/code/cpp/ir/implementation/aliased_ssa/internal/AliasedSSA.qll]: If `hasResultMemoryAccess` is fast, then it should be fine.

**Comment by jbj** [comment]
I think I remember Dave saying that no single predicate stands out as being slow(er). Is that right?

**Comment by dave-bartolomeo** [comment]
There are a couple SSA predicates that show up near the top on Wireshark. `SSAConstruction::DefUse::definitionReachesRank#ffff` in particular takes ~3x as long as the next predicate. However, that same predicate isn't even the top predicate on ChakraCore, where the timing distribution across the top 10 or so predicates is much more flat. I think that based on our discussion at the most recent stream meeting, this PR is ready to go, and we'll do more in-depth performance investigation on Wireshark once we've fixed or worked around enough of the invalid AST/IR problems that we can believe that our perf numbers are representative.

**Comment by jbj** [comment]
@dave-bartolomeo I've investigated the slowness in `definitionReachesRank` and filed https://jira.semmle.com/browse/CPP-390.

**Pull request merged**